fl_training.py

```diff
@@ -149,191 +194,169 @@ def compute_test_score(
 
+    # First, if only one node, fall back to dedicated single node function
     nodes_count = len(node_list)
-
-    # If only one node, fall back to dedicated single node function
     if nodes_count == 1:
         return compute_test_score_for_single_node(
             node_list[0], epoch_count, single_partner_test_mode, x_test, y_test
         )
 
-    # Else, follow a federated learning procedure
-    else:
-        print("\n## Training and evaluating model on multiple nodes: " + str(node_list))
-
-        model_list = [None] * nodes_count
-        epochs = epoch_count
-        score_matrix = np.zeros(shape=(epochs, nodes_count))
-        global_val_acc = []
-        global_val_loss = []
-
-        print("\n### Training model:")
-        for epoch in range(epochs):
-
-            print(
-                "\nEpoch #"
-                + str(epoch + 1)
-                + " out of "
-                + str(epochs)
-                + " total epochs"
-            )
-            is_first_epoch = epoch == 0
-            clear_session()
-
-            # Aggregation, intermediate evaluation and early stopping phase
-            if is_first_epoch:
-                # First epoch, no aggregation
-                print("   First epoch, generate model from scratch")
-
+    # Else, continue onto a federated learning procedure
+    print("\n## Training and evaluating model on multiple nodes: " + str(node_list))
+
+    # Initialize variables
+    model_list = [None] * nodes_count
+    epochs = epoch_count
+    score_matrix = np.zeros(shape=(epochs, nodes_count))
+    global_val_acc = []
+    global_val_loss = []
+    aggregation_weights = prepare_aggregation_weights(aggregation_weighting, nodes_count, node_list)
+
+    # Train model (iterate for each epoch and mini-batch)
+    print("\n### Training model:")
+    for epoch in range(epochs):
+
+        print(
+            "\nEpoch #"
+            + str(epoch + 1)
+            + " out of "
+            + str(epochs)
+            + " total epochs"
+        )
+        is_first_epoch = epoch == 0
+        clear_session()
+
+        # Split datasets in mini-batches
+        minibatched_x_train = [None] * nodes_count
+        minibatched_x_val = [None] * nodes_count
+        minibatched_y_train = [None] * nodes_count
+        minibatched_y_val = [None] * nodes_count
+        split_indices = np.arange(minibatch_count+1) / minibatch_count
+        for node_index, node in enumerate(node_list):
+            minibatched_x_train[node_index] = np.split(node.x_train, (split_indices * len(node.x_train)).astype(int))
+            minibatched_x_val[node_index] = np.split(node.x_val, (split_indices * len(node.x_val)).astype(int))
+            minibatched_y_train[node_index] = np.split(node.y_train, (split_indices * len(node.y_train)).astype(int))
+            minibatched_y_val[node_index] = np.split(node.y_val, (split_indices * len(node.y_val)).astype(int))
```

Comment:
A bit hard to follow.
I propose: `minibatched_x_train[node_index] = np.split(node.x_train, minibatches_count)`

Reply:
I did that in the beginning :-) But it doesn't work, `np.split()` can't accept an integer if it doesn't divide perfectly. So the proper way, as documented in NumPy doc, is to pass indices where to split.